src/zkevm_specs/evm/execution/log.py

```diff
@@ -0,0 +1,76 @@
+
+    # pop `mstart`, `msize` from stack
+    mstart = instruction.rlc_to_fq_exact(instruction.stack_pop(), 8)
+    msize = instruction.rlc_to_fq_exact(instruction.stack_pop(), 8)
+
+    # check not static call
+    instruction.constrain_equal(0, instruction.call_context_lookup(CallContextFieldTag.IsStatic))
+
+    # check contract_address in CallContext & TxLog
+    # use call context's  callee address as contract address
+    contract_address = instruction.call_context_lookup(CallContextFieldTag.CalleeAddress)
+    is_persistent = instruction.call_context_lookup(CallContextFieldTag.IsPersistent)
+    if not instruction.is_zero(is_persistent):
+        instruction.constrain_equal(
+            contract_address, instruction.tx_log_lookup(TxLogFieldTag.Address)
+        )
+
+    # constrain topics in stack & logs
+    is_topic_zeros = [1] * 4
+    topic_count = int(opcode) - Opcode.LOG0
+    for i in range(4):
+        if i < topic_count:
+            is_topic_zeros[i] = 0
+            topic = instruction.stack_pop()
+            if not instruction.is_zero(is_persistent):
+                instruction.constrain_equal(
+                    topic, instruction.tx_log_lookup(TxLogFieldTag.Topic, i)
+                )
+
+    # TOPIC_COUNT == Non zero topic count
+    assert sum(is_topic_zeros) == 4 - topic_count
+    # `is_topic_zeros` order must be from 0 --> 1
+    for i in range(1, 4):
+        diff = is_topic_zeros[i] - is_topic_zeros[i - 1]
+        instruction.constrain_bool(FQ(diff))
+
+    # check memory copy, should do in next step here
+    # When length != 0, constrain the state in the next execution state CopyToLog
+    if not instruction.is_zero(msize):
+        instruction.constrain_equal(instruction.next.execution_state, ExecutionState.CopyToLog)
+        next_aux = instruction.next.aux_data
+        instruction.constrain_equal(next_aux.src_addr, mstart)
+        instruction.constrain_equal(next_aux.src_addr_end, mstart + msize)
+        instruction.constrain_equal(next_aux.bytes_left, msize)
+        instruction.constrain_equal(next_aux.is_persistent, is_persistent)
+
+    # omit block number constraint even it is set within op code explicitly, because by default the circuit only handle
+    # current block, otherwise, block context lookup is required.
+    # calculate dynamic gas cost
+    next_memory_size, memory_expansion_gas = instruction.memory_expansion_dynamic_length(
+        mstart, msize
+    )
+    dynamic_gas = GAS_COST_LOG * (opcode - Opcode.LOG0) + 8 * msize + memory_expansion_gas
+
+    instruction.step_state_transition_in_same_context(
+        opcode,
+        rw_counter=Transition.delta(instruction.rw_counter_offset),
+        program_counter=Transition.delta(1),
+        stack_pointer=Transition.delta(2 + opcode - Opcode.LOG0),
+        state_write_counter=Transition.delta(1),
+        dynamic_gas_cost=dynamic_gas,
+        memory_size=Transition.to(next_memory_size),
+        log_id=Transition.delta(1),
```

Comment:
One last thing: log_id should only be `delta(1)` when `is_persistent = 1`.
Maybe you can do `log_id=Transition.delta(is_persistent)`

Reply:
Ah,  right , good catch!  will update it .

Reply:
updated in [c2f1e79](https://github.com/appliedzkp/zkevm-specs/pull/125/commits/c2f1e79faa999e32bd3c28a799c4ca69fb892815) as well as test!